### DISCUSSION TITLE
Internal API for custom extension

### DISCUSSION THREAD

**Discussion by alranel** [HASwitchPlate/openHASP#562] [General]
Hi! I'm trying to write a custom extension that syncs values with another remote service which does not use MQTT, but I'm having a hard time understanding what is the internal API that provides read/write access to the state of the openHASP widgets. Is it documented anywhere? :)

Additionally, it would be nice to load a JSONL file programmatically (for instance, auto-generating it based on some remote metadata). Is there an API also for that?

**Comment by fvanroie** [comment]
Once you get into custom coding, it is basically the Arduino and LVGL APIs that you need to use to access the system and objects.

**Reply by alranel** [reply]
I was thinking more about some higher level API such as `dispatch_text_line()` which seems promising :) Not sure it can be used to read properties of objects, but I'll cross fingers and try.
